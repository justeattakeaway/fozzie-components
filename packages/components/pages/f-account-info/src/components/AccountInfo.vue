<template>
    <f-card
        :card-heading="$t('accountDetails')"
        data-test-id="account-info"
        has-outline
        is-page-content-wrapper>
        <h2
            class="u-spacingBottom--large">
            {{ $t('yourDetails') }}
        </h2>

        <email-address-field :email-address="consumer.emailAddress" />

        <form
            method="post"
            @submit.prevent="onFormSubmit">
            <form-field
                v-model="consumer.firstName"
                maxlength="50"
                :label-text="$t('consumer.firstNameLabel')"
                :placeholder="$t('consumer.firstNamePlaceholder')"
                @blur="onBlur('firstName')">
                <template
                    v-if="$v.consumer.firstName.$invalid"
                    #error>
                    <f-error-message v-show="!$v.consumer.firstName.required && $v.consumer.firstName.$dirty">
                        {{ $t('validation.firstNameRequired') }}
                    </f-error-message>
                    <f-error-message v-show="!$v.consumer.firstName.isValidName">
                        {{ $t('validation.firstNameInvalid') }}
                    </f-error-message>
                </template>
            </form-field>

            <form-field
                v-model="consumer.lastName"
                maxlength="50"
                :label-text="$t('consumer.lastNameLabel')"
                :placeholder="$t('consumer.lastNamePlaceholder')"
                @blur="onBlur('lastName')">
                <template
                    v-if="$v.consumer.lastName.$invalid"
                    #error>
                    <f-error-message v-show="!$v.consumer.lastName.required && $v.consumer.lastName.$dirty">
                        {{ $t('validation.lastNameRequired') }}
                    </f-error-message>
                    <f-error-message v-show="!$v.consumer.lastName.isValidName">
                        {{ $t('validation.lastNameInvalid') }}
                    </f-error-message>
                </template>
            </form-field>

            <form-field
                v-model="consumer.phoneNumber"
                maxlength="16"
                :label-text="$t('consumer.phoneNumberLabel')"
                :placeholder="$t('consumer.phoneNumberPlaceholder')"
                @blur="onBlur('phoneNumber')">
                <template
                    v-if="$v.consumer.phoneNumber.$invalid"
                    #error>
                    <f-error-message v-show="!$v.consumer.phoneNumber.required && $v.consumer.phoneNumber.$dirty">
                        {{ $t('validation.phoneNumberRequired') }}
                    </f-error-message>
                    <f-error-message v-show="!$v.consumer.phoneNumber.isValidPhoneNumber && $v.consumer.phoneNumber.required && $v.consumer.phoneNumber.$dirty">
                        {{ $t('validation.phoneNumberInvalid') }}
                    </f-error-message>
                </template>
            </form-field>

            <h2
                class="u-spacingBottom--large">
                {{ $t('deliveryAddress') }}
            </h2>

            <form-field
                v-model="consumer.line1"
                maxlength="50"
                :label-text="$t('consumer.addressLabel')"
                :placeholder="$t('consumer.line1Placeholder')"
                @blur="onBlur('line1')">
                <template
                    v-if="$v.consumer.line1.$invalid"
                    #error>
                    <f-error-message v-show="!$v.consumer.line1.required && $v.consumer.line1.$dirty">
                        {{ $t('validation.line1Required') }}
                    </f-error-message>
                </template>
            </form-field>

            <form-field
                v-model="consumer.line2"
                maxlength="50"
                :placeholder="$t('consumer.line2Placeholder')" />

            <form-field
                v-model="consumer.line3"
                maxlength="50"
                :placeholder="$t('consumer.line3Placeholder')" />

            <form-field
                v-model="consumer.locality"
                maxlength="50"
                :label-text="$t('consumer.localityLabel')"
                :placeholder="$t('consumer.localityPlaceholder')"
                @blur="onBlur('locality')">
                <template
                    v-if="$v.consumer.locality.$invalid"
                    #error>
                    <f-error-message v-show="!$v.consumer.locality.required && $v.consumer.locality.$dirty">
                        {{ $t('validation.localityRequired') }}
                    </f-error-message>
                </template>
            </form-field>

            <form-field
                v-model="consumer.postcode"
                maxlength="50"
                :label-text="$t('consumer.postcodeLabel')"
                :placeholder="$t('consumer.postcodePlaceholder')"
                @blur="onBlur('postcode')">
                <template
                    v-if="$v.consumer.postcode.$invalid"
                    #error>
                    <f-error-message v-show="!$v.consumer.postcode.required && $v.consumer.postcode.$dirty">
                        {{ $t('validation.postcodeRequired') }}
                    </f-error-message>
                    <f-error-message v-show="!$v.consumer.postcode.isValidPostcode && $v.consumer.postcode.required">
                        {{ $t('validation.postcodeInvalid') }}
                    </f-error-message>
                </template>
            </form-field>

            <f-button
                :class="[$style['c-accountInfo-submitButton']]"
                data-test-id="account-info-submit-button"
                button-type="primary"
                button-size="large"
                is-full-width
                action-type="submit"
                :is-loading="isFormSubmitting">
                {{ $t('buttons.saveChanges') }}
            </f-button>
        </form>

        <f-button
            :class="[$style['c-accountInfo-changePasswordButton']]"
            data-test-id="account-info-submit-button"
            button-type="secondary"
            href="/change-password?returnurl=/account/info"
            button-size="large"
            is-full-width
            action-type="submit">
            {{ $t('buttons.changePassword') }}
        </f-button>

        <delete-account />
    </f-card>
</template>

<script>
import { mapActions, mapState } from 'vuex';
import { VueGlobalisationMixin } from '@justeat/f-globalisation';
import FErrorMessage from '@justeat/f-error-message';
import '@justeat/f-error-message/dist/f-error-message.css';
import FCard from '@justeat/f-card';
import '@justeat/f-card/dist/f-card.css';
import FormField from '@justeat/f-form-field';
import '@justeat/f-form-field/dist/f-form-field.css';
import FButton from '@justeat/f-button';
import '@justeat/f-button/dist/f-button.css';
import EmailAddressField from './EmailAddressField.vue';
import DeleteAccount from './DeleteAccount.vue';
import AccountInfoValidationMixin from './AccountInfoValidationMixin.vue';
import tenantConfigs from '../tenants';
import ConsumerApi from '../services/providers/Consumer.api';
import fAccountInfoModule from '../store/accountInfo.module';
import {
    EVENT_SPINNER_STOP_LOADING
} from '../constants';

export default {
    components: {
        FCard,
        FormField,
        FButton,
        FErrorMessage,
        EmailAddressField,
        DeleteAccount
    },

    mixins: [
        VueGlobalisationMixin,
        AccountInfoValidationMixin
    ],

    props: {
        authToken: {
            type: String,
            default: null
        },
        isAuthFinished: {
            type: Boolean,
            required: true
        },
        smartGatewayBaseUrl: {
            type: String,
            required: true
        }
    },

    data () {
        return {
            consumerApi: new ConsumerApi({
                httpClient: this.$http,
                cookies: this.$cookies,
                baseUrl: this.smartGatewayBaseUrl
            }),
            tenantConfigs,
            isFormSubmitting: false
        };
    },

    computed: {
        ...mapState('fAccountInfoModule', [
            'consumer'
        ])
    },

    watch: {
        isAuthFinished: {
            immediate: true,
            async handler (value) {
                if (value) {
                    await this.initialise();
                }
            }
        }
    },

    created () {
        if (!this.$store.hasModule('fAccountInfoModule')) {
            this.$store.registerModule('fAccountInfoModule', fAccountInfoModule);
        }
    },

    async mounted () {
        if (this.isAuthFinished) {
            await this.initialise();
        }
    },

    methods: {
        ...mapActions('fAccountInfoModule', [
            'loadConsumerDetails',
            'editConsumerDetails'
        ]),

        async initialise () {
            try {
                await this.loadConsumerDetails({ api: this.consumerApi, authToken: this.authToken });
            } catch (error) {
                // TODO - to be added with next ticket
            } finally {
                this.$nextTick(() => {
                    this.$parent.$emit(EVENT_SPINNER_STOP_LOADING);
                });
            }
        },

        onFormSubmit () {
            if (this.isFormInvalid()) {
                this.logValidationFailure();
                return;
            }

            this.setSubmittingState(true);

            try {
                // TODO - to be added with next ticket
                this.$log.info('Submitted Form', ['account-info', 'account-pages']);
            } catch (error) {
                // TODO - to be added with next ticket
            } finally {
                this.setSubmittingState(false);
            }
        },

        /**
        * Sets the flag to inform the Template of whether the form is currently submitting or not
        * @param {boolean} isFormSubmitting - True = Form is being submitted / False = Form is not being submitted
        */
        setSubmittingState (isFormSubmitting) {
            this.isFormSubmitting = isFormSubmitting;
        }
    }
};
</script>

<style lang="scss" module>
.c-accountInfo-submitButton {
    margin-top: spacing(x4);
}

.c-accountInfo-changePasswordButton {
    margin-top: spacing(x2);
}
</style>
