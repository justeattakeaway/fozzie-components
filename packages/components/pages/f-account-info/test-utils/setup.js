import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import { VueI18n } from '@justeat/f-globalisation';
import tenantConfigs from '../src/tenants';

export const localVue = createLocalVue();
localVue.use(VueI18n);
localVue.use(Vuex);

export const i18n = {
    locale: 'en-GB',
    messages: {
        'en-GB': tenantConfigs['en-GB'].messages
    }
};

export const baseUrl = 'https://smartGatewayBaseUrl.com';

export const token = 'some-auth-token';

export const conversationId = 'b7386108-95e6-4e73-9421-5b066c089153';

export const consumerDetailsGetResponse = {
    data: {
        ConsumerId: 1050450174,
        Id: '8It2lEaJjG82nJtxO00Iab0CoF0=',
        CreatedDate: '2019-04-16T17:25:00.000Z',
        EmailAddress: 'satchin.moriyoka@example.com',
        FirstName: 'Satchin',
        LastName: 'Moriyoka',
        PhoneNumber: '+441234567890',
        DateOfBirth: '1914-04-16T17:25:00.000Z',
        TermsAndConditionsId: 0,
        WantsNewsletter: true,
        AccountType: 'Default',
        ConsumerStatus: 'ReturnConsumer'
    }
};

const consumerAddressGetResponse = {
    AddressId: 1050450174,
    City: 'Bristol',
    ZipCode: 'BS1 4DJ',
    AddressName: 'Office',
    IsDefault: true, // Key field
    Line1: 'Just Eat',
    Line2: 'Broad Quay House'
};

export const consumerAddressesGetResponse = {
    data: [{
        ...consumerAddressGetResponse
    },
    {
        AddressId: 1050450175,
        City: 'Bristol',
        ZipCode: 'BS1 9DJ',
        AddressName: 'Office',
        IsDefault: false,
        Line1: 'Takeaway',
        Line2: 'Broad Quay House'
    }]
};

export const consumerDetailsMappedModel = {
    id: consumerDetailsGetResponse.data.ConsumerId,
    firstName: consumerDetailsGetResponse.data.FirstName,
    lastName: consumerDetailsGetResponse.data.LastName,
    phoneNumber: consumerDetailsGetResponse.data.PhoneNumber,
    emailAddress: consumerDetailsGetResponse.data.EmailAddress
};

export const consumerAddressMappedModel = {
    line1: consumerAddressGetResponse.Line1,
    line2: consumerAddressGetResponse.Line2,
    line3: consumerAddressGetResponse.Line3,
    locality: consumerAddressGetResponse.City,
    postcode: consumerAddressGetResponse.ZipCode
};

export const consumerViewModel = {
    ...consumerDetailsMappedModel,
    ...consumerAddressMappedModel
};

export const consumerStateModel = {
    consumer: {
        ...consumerDetailsMappedModel,
        ...consumerAddressMappedModel
    }
};
