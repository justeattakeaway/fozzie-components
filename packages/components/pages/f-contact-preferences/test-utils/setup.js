export const contactPreferencesGetResponse = {
    data: {
        Preferences: [{
            DisplayName: 'Order status',
            Sort: 1,
            Key: 'orderstatus',
            Push: false,
            Email: true,
            Sms: true
        }, {
            DisplayName: 'Review meal',
            Sort: 2,
            Key: 'reviewmeal',
            Push: false,
            Email: false,
            Sms: false
        }, {
            DisplayName: 'News & offers',
            Sort: 3,
            Key: 'news',
            Push: false,
            Email: false,
            Sms: false
        }
        ],
        PreferenceVersionViewed: 0
    }
};

export const contactPreferencesViewModel = {
    preferenceVersionViewed: 0,
    preferences: [{
        displayName: 'Order status',
        isEmailEnabled: false,
        emailValue: true,
        key: 'orderstatus',
        pushEnabled: false,
        pushValue: false,
        isSmsEnabled: false,
        smsValue: true,
        sort: 1,
        visible: true
    }, {
        displayName: 'Review meal',
        isEmailEnabled: true,
        emailValue: false,
        key: 'reviewmeal',
        pushEnabled: true,
        pushValue: false,
        isSmsEnabled: true,
        smsValue: false,
        sort: 2,
        visible: false
    }, {
        displayName: 'News & offers',
        isEmailEnabled: true,
        emailValue: false,
        key: 'news',
        pushEnabled: true,
        pushValue: false,
        isSmsEnabled: true,
        smsValue: false,
        sort: 3,
        visible: true
    }]
};

export const filteredContactPreferencesModel = [{
    displayName: 'Order status',
    isEmailEnabled: false,
    emailValue: true,
    key: 'orderstatus',
    pushEnabled: false,
    pushValue: false,
    isSmsEnabled: false,
    smsValue: true,
    sort: 1,
    visible: true
}, {
    displayName: 'News & offers',
    isEmailEnabled: true,
    emailValue: false,
    key: 'news',
    pushEnabled: true,
    pushValue: false,
    isSmsEnabled: true,
    smsValue: false,
    sort: 3,
    visible: true
}];

export const contactPreferencesUpdateModel = {
    Preferences: [{
        DisplayName: 'Order status',
        Sort: 1,
        Key: 'orderstatus',
        Push: false,
        Email: true,
        Sms: true
    }, {
        DisplayName: 'Review meal',
        Sort: 2,
        Key: 'reviewmeal',
        Push: false,
        Email: false,
        Sms: false
    }, {
        DisplayName: 'News & offers',
        Sort: 3,
        Key: 'news',
        Push: false,
        Email: false,
        Sms: false
    }],
    DeviceToken: null,
    DeviceType: null,
    PhoneNumber: null,
    PreferenceVersionViewed: 0
};

export const baseUrl = 'https://smartGatewayBaseUrl.com';

export const token = 'some-auth-token';

export const conversationId = 'b7386108-95e6-4e73-9421-5b066c089153';

export const contactPreferencesUpdateBody = {
    Preference: [{
        DisplayName: 'Order status',
        Sort: 1,
        Key: 'orderstatus',
        Push: false,
        Email: true,
        Sms: true
    }, {
        DisplayName: 'Review meal',
        Sort: 2,
        Key: 'reviewmeal',
        Push: false,
        Email: false,
        Sms: false
    }, {
        DisplayName: 'News & offers',
        Sort: 3,
        Key: 'news',
        Push: false,
        Email: true,
        Sms: false
    }
    ],
    DeviceToken: null,
    DeviceType: null,
    PhoneNumber: null,
    PreferenceVersionViewed: 0
};
