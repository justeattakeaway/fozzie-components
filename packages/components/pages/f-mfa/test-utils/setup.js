export const baseUrl = 'https://smartGatewayBaseUrl.com';

export const conversationId = 'b7386108-95e6-4e73-9421-5b066c089153';

// eslint-disable-next-line camelcase
export const postBody = { mfa_token: 'some_mfa_token', otp: 'ABC123' };
