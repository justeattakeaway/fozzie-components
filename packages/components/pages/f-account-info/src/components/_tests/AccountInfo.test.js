import Vuex from 'vuex';
import { shallowMount } from '@vue/test-utils';
import AccountInfo from '../AccountInfo.vue';
import {
    localVue,
    i18n,
    baseUrl,
    token,
    consumerStateModel
} from '../../../test-utils/setup';

let wrapper;
let cookiesSpy;
let httpSpy;
let sutMocks;
let sutProps;
let dataDefaults;
let registerStoreModuleSpy;
let initialiseSpy;

const storeActions = {
    loadConsumerDetails: jest.fn(),
    editConsumerDetails: jest.fn()
};

const storeState = consumerStateModel;

const createStore = ({ state, actions }) => new Vuex.Store({
    modules: {
        fAccountInfoModule: {
            state,
            actions,
            namespaced: true
        }
    }
});

const mountAccountInfo = ({
    actions = storeActions,
    state = storeState,
    mocks = sutMocks,
    propsData = sutProps,
    data = dataDefaults,
    storeOverride = null,
    initialiseOverride = null
} = {}) => {
    const store = storeOverride || createStore({ state, actions });
    store.registerModule = registerStoreModuleSpy;

    if (initialiseOverride) {
        AccountInfo.methods.initialise = initialiseOverride;
    }

    const mock = shallowMount(AccountInfo, {
        i18n,
        localVue,
        propsData,
        data,
        store,
        mocks
    });

    return mock;
};

describe('AccountInfo', () => {
    beforeEach(() => {
        // Arrange
        registerStoreModuleSpy = jest.fn();
        dataDefaults = () => ({
            isFormDirty: false,
            shouldShowErrorPage: false
        });
        cookiesSpy = jest.fn();
        httpSpy = jest.fn();
        sutMocks = {
            $parent: {
                $emit: jest.fn()
            },
            $http: httpSpy,
            $cookies: cookiesSpy
        };
        sutProps = {
            authToken: token,
            isAuthFinished: true,
            smartGatewayBaseUrl: baseUrl
        };
    });

    afterEach(() => {
        initialiseSpy = null;
        jest.clearAllMocks();
    });

    describe('when mounting the component', () => {
        it('should not call initialise() method if the authorisation has not completed', () => {
            // Arrange
            initialiseSpy = jest.fn();
            sutProps = { ...sutProps, isAuthFinished: false };

            // Act
            wrapper = mountAccountInfo({ initialiseOverride: initialiseSpy });

            // Assert
            expect(initialiseSpy).not.toHaveBeenCalled();
        });

        it('should only call initialise() method once the authorisation has completed', async () => {
            // Arrange
            initialiseSpy = jest.fn();
            sutProps = { ...sutProps, isAuthFinished: false };

            // Act
            wrapper = mountAccountInfo({ initialiseOverride: initialiseSpy });

            // Assert
            expect(initialiseSpy).not.toHaveBeenCalled();

            // Act 2
            await wrapper.setProps({ isAuthFinished: true });

            // Assert 2
            expect(initialiseSpy).toHaveBeenCalled();
        });
    });
});
