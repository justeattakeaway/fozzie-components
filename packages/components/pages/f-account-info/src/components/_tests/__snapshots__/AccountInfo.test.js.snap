// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountInfo \`methods\` onFormSubmit :: publishing analytics form is valid address has changed 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "action": "save_accountinfo_changes",
        "category": "account",
        "event": "trackEvent",
        "label": "submit",
      },
    ],
    Array [
      Object {
        "action": "account info",
        "category": "my account",
        "event": "trackEvent",
        "label": "address change intent",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`AccountInfo \`methods\` onFormSubmit :: publishing analytics form is valid address has not changed 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "action": "save_accountinfo_changes",
        "category": "account",
        "event": "trackEvent",
        "label": "submit",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
